src/Nethermind/Nethermind.Blockchain/Bloom/BloomStorage.cs

```diff
@@ -0,0 +1,374 @@
+                {
+                    var levelBloom = levelBlooms[index];
+                    levelBloom.Bloom.Accrue(blockHeaderBloom);
+                    
+                    if (i % levelBloom.Level.LevelElementSize == 0)
+                    {
+                        levelBloom.Level.StoreMigration(blockHeader.Number, levelBloom.Bloom);
+                        levelBlooms[index] = (levelBloom.Level, new Core.Bloom());
+
+                        if (levelBloom.Level.LevelElementSize == batchSize)
+                        {
+                            MigratedBlockNumber += batchSize;
+                            Set(MigrationBlockNumberKey, MigratedBlockNumber);
+                        }
+                    }
+                }
+
+                lastBlockNumber = blockHeader.Number;
+            }
+
+            if (i % batchSize != 0)
+            {
+                for (var index = 0; index < levelBlooms.Length; index++)
+                {
+                    var levelBloom = levelBlooms[index];
+                    levelBloom.Level.Store(lastBlockNumber, levelBloom.Bloom);
+                }
+                
+                MigratedBlockNumber += i;
+                Set(MigrationBlockNumberKey, MigratedBlockNumber);
+            }
+            
+            if (MigratedBlockNumber >= MinBlockNumber - 1)
+            {
+                MinBlockNumber = 0;
+                Set(MinBlockNumberKey, MinBlockNumber);
+            }
+        }
+
+        private void Set(Keccak key, long value)
+        {
+            _bloomInfoDb.Set(key, value.ToBigEndianByteArrayWithoutLeadingZeros());
+        }
+
+        public void Dispose()
+        {
+            for (int i = 0; i < _storageLevels.Length; i++)
+            {
+                _storageLevels[i].Dispose();
+            }
+        }
+
+        public IBloomEnumeration GetBlooms(long fromBlock, long toBlock) => new BloomEnumeration(_storageLevels, Math.Max(fromBlock, MinBlockNumber), Math.Min(toBlock, MaxBlockNumber));
+
+        private class BloomStorageLevel : IDisposable
+        {
+            public readonly byte Level;
+            public readonly int LevelElementSize;
+            public readonly Average Average = new Average();
+            
+            private readonly IFileStore _fileStore;
+            private readonly bool _migrationStatistics;
+            private readonly LruCache<long, Core.Bloom> _cache;
+            private readonly byte[] _bytes = new byte[Core.Bloom.ByteLength];
+
+            public BloomStorageLevel(IFileStore fileStore, in byte level, in int levelElementSize, in int levelMultiplier, bool migrationStatistics)
+            {
+                _fileStore = fileStore;
+                Level = level;
+                LevelElementSize = levelElementSize;
+                _migrationStatistics = migrationStatistics;
+                _cache = new LruCache<long, Core.Bloom>(levelMultiplier);
```

Comment:
maybe better to cache the coarser level multiplier elements
so if you have 4, 256, then you will cache 256 on level 4, and some default for level 256

Reply:
or better if you have
0 -> 4
1 -> 256
2 -> 256 * 16

then cache sizes are:
default * 256/4
default * 256*16 / 256
default 
----
default can be 4 or 16

Reply:
Not needed, this cache is currently used more for inserting, when reading we ignore them ang just read sequentially from flat file, which is faster. For inserting we rarely insert in long past. Maybe we could drop them completely.